File: python/2023/day02.py
from __future__ import annotations


class Subset(tuple):
    def __new__(cls, *args) -> Subset:
        return tuple.__new__(cls, args)

    def parse(s: str) -> Subset:
        r, g, b = 0, 0, 0

        for splt in s.split(","):
            match splt.split():
                case [num, "red"]: r = int(num)
                case [num, "green"]: g = int(num)
                case [num, "blue"]: b = int(num)

        return Subset(r, g, b)

    @property
    def red(self):
        return self[0]

    @property
    def green(self):
        return self[1]

    @property
    def blue(self):
        return self[2]


class Game:
    id: int
    subsets: list[Subset]

    def __init__(self, line: str):
        splt = line.split(":")
        self.id = int(splt[0].split()[-1])
        self.subsets = [Subset.parse(s) for s in splt[1].split(";")]


def part1(input: str) -> int:
    lim = Subset(12, 13, 14)
    answer = 0

    for game in [Game(line) for line in input.splitlines()]:
        for ss in game.subsets:
            if ss.red > lim.red or ss.green > lim.green or ss.blue > lim.blue:
                break
        else:
            answer += game.id

    return answer


def part2(input: str) -> int:
    answer = 0

    for game in [Game(line) for line in input.splitlines()]:
        red = max(map(lambda ss: ss.red, game.subsets))
        green = max(map(lambda ss: ss.green, game.subsets))
        blue = max(map(lambda ss: ss.blue, game.subsets))
        answer += red * green * blue

    return answer


TEST_INPUT = """Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green"""

PART1_TESTS = [
    (TEST_INPUT, 8),
]

PART2_TESTS = [
    (TEST_INPUT, 2286),
]

File: rust/2023/src/bin/day02.rs
#[derive(Clone)]
struct Game {
    red: u32,
    green: u32,
    blue: u32,
}

impl TryFrom<&str> for Game {
    type Error = Error;

    fn try_from(line: &str) -> Result<Self, Self::Error> {
        let (_, line) = line
            .split_once(':')
            .ok_or(anyhow!("failed to split line"))?;

        let mut game = Self::new();

        for (num, col) in line.split_whitespace().tuples() {
            match col.chars().next() {
                Some('r') => game.red = game.red.max(num.parse()?),
                Some('g') => game.green = game.green.max(num.parse()?),
                Some('b') => game.blue = game.blue.max(num.parse()?),
                _ => return Err(anyhow!("failed to parse as colour: {col}")),
            }
        }

        Ok(game)
    }
}

impl Game {
    fn new() -> Self {
        Self {
            red: 0,
            green: 0,
            blue: 0,
        }
    }
}

#[derive(Clone)]
struct Solution {
    games: Vec<Game>,
}

impl Solver for Solution {
    fn new(input: &str) -> Anyhow<Self> {
        Ok(Self {
            games: input
                .lines()
                .map(Game::try_from)
                .collect::<Anyhow<Vec<_>>>()?,
        })
    }

    fn part1(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self.games.iter().enumerate().fold(0, |acc, (i, game)| {
            if game.red > 12 || game.green > 13 || game.blue > 14 {
                acc
            } else {
                acc + i + 1
            }
        }))
    }

    fn part2(&mut self) -> Anyhow<impl fmt::Display> {
        Ok(self
            .games
            .iter()
            .fold(0, |acc, game| acc + game.red * game.green * game.blue))
    }
}

aoc::solution!();

#[cfg(test)]
mod test {
    use super::{Solution, Solver};

    const INPUT: &str = r"Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green";

    #[test]
    fn test_part1() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part1().unwrap().to_string();
        assert_eq!(answer, "8");
    }

    #[test]
    fn test_part2() {
        let mut solution = Solution::new(INPUT).unwrap();
        let answer = solution.part2().unwrap().to_string();
        assert_eq!(answer, "2286");
    }
}
